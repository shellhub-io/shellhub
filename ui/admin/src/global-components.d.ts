// src/global-components.d.ts
import SnackbarComponent from "./components/Snackbar/Snackbar.vue";

declare module "@vue/runtime-core" {
  export interface GlobalComponents {
    SnackbarComponent: typeof SnackbarComponent;
    RouterLink: typeof import("vue-router")["RouterLink"];
    RouterView: typeof import("vue-router")["RouterView"];
  }
}
