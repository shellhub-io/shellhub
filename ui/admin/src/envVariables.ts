/* eslint-disable import/prefer-default-export */
export const envVariables = {
  announcementsEnable: process.env.SHELLHUB_ANNOUNCEMENTS === "true",
  tinyMceKey: process.env.TINYMCE_API_KEY,
};
