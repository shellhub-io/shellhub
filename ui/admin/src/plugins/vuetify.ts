// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";
import { VNumberInput } from "vuetify/labs/VNumberInput";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
  components: {
    VNumberInput,
  },
});
