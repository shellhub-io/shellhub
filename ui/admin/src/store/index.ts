import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";
import { stats, StatsState } from "./modules/stats";

import { UsersState, users } from "./modules/users";
import { AuthState, auth } from "./modules/auth";
import { DevicesState, devices } from "./modules/devices";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { layout, LayoutState } from "./modules/layout";
import { license, LicenseState } from "./modules/license";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { sessions, SessionsState } from "./modules/sessions";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { spinner, SpinnerState } from "./modules/spinner";
import { announcement, AnnouncementState } from "./modules/announcement";
import { instance, InstanceState } from "./modules/instance";

export interface State {
  user: UsersState;
  auth: AuthState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  license: LicenseState;
  namespaces: NamespacesState;
  publicKeys: PublicKeysState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  announcement: AnnouncementState;
  instance: InstanceState;
}

// eslint-disable-next-line symbol-description
export const key: InjectionKey<Store<State>> = Symbol();

export const store = createStore<State>({
  modules: {
    users,
    auth,
    devices,
    firewallRules,
    layout,
    license,
    namespaces,
    publicKeys,
    sessions,
    snackbar,
    spinner,
    stats,
    announcement,
    instance,
  },
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
