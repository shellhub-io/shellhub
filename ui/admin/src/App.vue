<template>
  <v-app>
    <component :is="layout" :data-test="`${layout}-component`" />
  </v-app>
</template>

<script lang="ts">
import { defineComponent, onMounted, computed } from "vue";
import { useRouter } from "vue-router";
import SimpleLayout from "./layouts/SimpleLayout.vue";
import AppLayout from "./layouts/AppLayout.vue";
import { useStore } from "./store";
import { INotificationsError } from "./interfaces/INotifications";

export default defineComponent({
  name: "App",
  components: {
    appLayout: AppLayout,
    simpleLayout: SimpleLayout,
  },
  setup() {
    const store = useStore();
    const router = useRouter();

    const layout = computed(() => store.getters["layout/getLayout"]);

    const isLoggedIn = computed(() => store.getters["auth/isLoggedIn"]);

    const currentRoute = computed(() => router.currentRoute.value.path);

    onMounted(async () => {
      if (!isLoggedIn.value) {
        try {
          await store.dispatch("auth/logout");
          store.dispatch("layout/setLayout", "simpleLayout");
          router.push("/login");
        } catch {
          store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.namespaceLoad);
        }
      }

      if (isLoggedIn.value && currentRoute.value !== "/login") {
        const license = await store.dispatch("license/get");
        if (!license || license.expired) {
          store.dispatch("snackbar/showSnackbarErrorAction", INotificationsError.license);
          store.dispatch("layout/setLayout", "appLayout");
          router.push("/license");
        }
        store.dispatch("layout/setLayout", "appLayout");
      }
    });

    return {
      layout,
    };
  },
});
</script>
