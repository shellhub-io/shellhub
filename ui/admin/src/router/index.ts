import { createRouter, createWebHistory } from "vue-router";
import Login from "@admin/views/Login.vue";
import Dashboard from "@admin/views/Dashboard.vue";
import Users from "@admin/views/Users.vue";
import SettingsLicense from "@admin/components/Settings/SettingsLicense.vue";
import SettingsAuthentication from "@admin/components/Settings/SettingsAuthentication.vue";
import Namespaces from "@admin/views/Namespaces.vue";
import Settings from "@admin/views/Settings.vue";

import { store } from "../store";

const routes = [
  {
    path: "/login",
    name: "login",
    component: Login,
  },
  {
    path: "/",
    name: "dashboard",
    component: Dashboard,
  },
  {
    path: "/users",
    name: "users",
    component: Users,
  },
  {
    path: "/user/:id",
    name: "userDetails",
    component: () => import("@admin/views/UserDetails.vue"),
  },
  {
    path: "/devices",
    name: "devices",
    component: () => import("@admin/views/Device.vue"),
    redirect: {
      name: "listDevices",
    },
    children: [
      {
        path: "",
        name: "listDevices",
        component: () => import("@admin/views/Device.vue"),
      },
    ],
  },
  {
    path: "/device/:id",
    name: "deviceDetails",
    component: () => import("@admin/views/DeviceDetails.vue"),
  },
  {
    path: "/sessions",
    name: "sessions",
    component: () => import("@admin/views/Sessions.vue"),
  },
  {
    path: "/session/:id",
    name: "sessionDetails",
    component: () => import("@admin/views/SessionDetails.vue"),
  },
  {
    path: "/settings",
    name: "Settings",
    component: Settings,
    redirect: { name: "SettingProfile" },
    children: [
      {
        path: "authentication",
        name: "SettingAuthentication",
        component: SettingsAuthentication,
      },
      {
        path: "license",
        name: "SettingLicense",
        component: SettingsLicense,
      },
    ],
  },
  {
    path: "/firewall-rules",
    name: "firewall-rules",
    component: () => import("@admin/views/FirewallRules.vue"),
  },
  {
    path: "/firewall-rules/:id",
    name: "firewallRulesDetails",
    component: () => import("@admin/views/FirewallRulesDetails.vue"),
  },
  {
    path: "/namespaces",
    name: "namespaces",
    component: Namespaces,
  },
  {
    path: "/namespace/:id",
    name: "namespaceDetails",
    component: () => import("@admin/views/NamespaceDetails.vue"),
  },
  {
    path: "/announcements",
    name: "announcements",
    component: () => import("@admin/views/Announcements.vue"),
  },
  {
    path: "/announcement/:uuid",
    name: "announcementDetails",
    component: () => import("@admin/views/AnnouncementDetails.vue"),
  },
  {
    path: "/new-announcement",
    name: "new-announcement",
    component: () => import("@admin/views/NewAnnouncement.vue"),
  },
];

const router = createRouter({
  history: createWebHistory("/admin/"),
  routes,
});

router.beforeEach((to, from, next) => {
  if (to.path !== "/login") {
    if (store.getters["auth/isLoggedIn"]) {
      return next();
    }
    return next(`/login?redirect=${to.path}`);
  }
  if (store.getters["auth/isLoggedIn"]) {
    if (to.path === "/login" && to.query.token) {
      return next();
    }
    return next("/");
  }

  return next();
});

export default router;
