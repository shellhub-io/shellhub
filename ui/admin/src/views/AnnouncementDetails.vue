<template>
  <div class="d-flex pa-0 align-center">
    <h1>Announcement Details</h1>
  </div>

  <v-card class="mt-2 pa-4">
    <v-card-text>
      <div>
        <div class="text-overline mt-3">
          <h3>Uuid:</h3>
        </div>
        <div :data-test="announcement.uuid">
          <p>{{ announcement.uuid }}</p>
        </div>
      </div>

      <div>
        <div class="text-overline mt-3">
          <h3>Title:</h3>
        </div>
        <div :data-test="announcement.title">
          <p>{{ announcement.title }}</p>
        </div>
      </div>

      <div>
        <div class="text-overline mt-3">
          <h3>Content:</h3>
        </div>
        <div class="pa-2">
          <div :data-test="announcement.content">
            <div class="content-announcement" v-html="contentToHtml" />
          </div>
        </div>
      </div>

      <div>
        <div class="text-overline mt-3">
          <h3>Date:</h3>
        </div>
        <div :data-test="announcement.date">
          <p>{{ date }}</p>
        </div>
      </div>
    </v-card-text>
  </v-card>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref } from "vue";
import { useRoute } from "vue-router";
import MarkdownIt from "markdown-it";
import moment from "moment";
import { useStore } from "../store";

export default defineComponent({
  name: "Announcement",
  setup() {
    const md = new MarkdownIt();
    const store = useStore();
    const route = useRoute();
    const announcementId = computed(() => route.params.uuid);
    const announcement = computed(
      () => store.getters["announcement/announcement"],
    );
    const contentToHtml = ref("");
    const date = ref("");

    onMounted(async () => {
      await store.dispatch(
        "announcement/getAnnouncement",
        announcementId.value,
      );

      if (announcement.value) {
        contentToHtml.value = md.render(announcement.value.content);
        date.value = moment(announcement.value.date).format("LL");
      }
    });

    return {
      announcement,
      contentToHtml,
      date,
    };
  },
});
</script>

<style lang="scss">
.content-announcement {
  p, span, div, h1, h2, h3, h4, h5, h6 {
    margin: .5rem;
    line-height: 1.4;
  }

  ul, ol {
    padding: .5rem 1.5rem;
  }

  img {
    max-width: 100%;
  }

  blockquote {
    display: block;
    margin-block-start: 1em;
    margin-block-end: 1em;
    margin-inline-start: 40px;
    margin-inline-end: 40px;
  }
}
</style>
