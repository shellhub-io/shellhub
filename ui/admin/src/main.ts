import { createApp } from "vue";
import vuetify from "@ui/plugins/vuetify";
import App from "./App.vue";
import { loadFonts } from "./plugins/webfontloader";
import { key, store } from "./store";
import router from "./router";
import SnackbarComponent from "./components/Snackbar/Snackbar.vue";

const app = createApp(App);

loadFonts();

app.use(vuetify);
app.use(router);
app.use(store, key);
app.component("SnackbarComponent", SnackbarComponent);
app.mount("#app");
