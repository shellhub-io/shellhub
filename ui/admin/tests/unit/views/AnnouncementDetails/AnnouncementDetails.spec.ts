import { createStore } from "vuex";
import { createVuetify } from "vuetify";
import { mount, VueWrapper } from "@vue/test-utils";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { key } from "../../../../src/store";
import routes from "../../../../src/router";
import AnnouncementDetails from "../../../../src/views/AnnouncementDetails.vue";

type AnnouncementDetailsWrapper = VueWrapper<InstanceType<typeof AnnouncementDetails>>;

const announcementDetail = {
  uuid: "eac7e18d-7127-41ca-b68b-8242dfdbaf4c",
  title: "Announcement 1",
  content: "## ShellHub new features \n - New feature 1 \n - New feature 2 \n - New feature 3",
  date: "2022-12-15T19:45:45.618Z",
};

const announcementContentInHtml = `<h2>ShellHub new features</h2>
<ul>
<li>New feature 1</li>
<li>New feature 2</li>
<li>New feature 3</li>
</ul>
`;

const mockRoute = {
  params: {
    uuid: "eac7e18d-7127-41ca-b68b-8242dfdbaf4c",
  },
};

describe("Announcement Details", () => {
  const store = createStore({
    state: {
      announcement: announcementDetail,
    },
    getters: {
      "announcement/announcement": () => announcementDetail,
    },
    actions: {
      "announcement/getAnnouncement": vi.fn(),
      "announcement/announcements": vi.fn(),
    },
  });

  const vuetify = createVuetify();

  let wrapper: AnnouncementDetailsWrapper;

  beforeEach(() => {
    wrapper = mount(AnnouncementDetails, {
      global: {
        plugins: [[store, key], vuetify, routes],
        mocks: {
          $route: mockRoute,
          $router: {
            push: vi.fn(),
          },
        },
      },
    });
  });

  it("Is a Vue instance", () => {
    expect(wrapper).toBeTruthy();
  });

  it("Renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Has the correct data", async () => {
    expect(wrapper.vm.announcement).toEqual(announcementDetail);
  });

  it("Render the correct title", () => {
    expect(wrapper.find("h1").text()).toEqual("Announcement Details");
  });

  it("Should render the props of the announcement in the Screen", () => {
    expect(wrapper.find(`[data-test='${announcementDetail.uuid}']`).text()).toContain(announcementDetail.uuid);
    expect(wrapper.find(`[data-test='${announcementDetail.title}']`).text()).toContain(announcementDetail.title);
  });

  it("Shoud render the markdown content in HTML", () => {
    expect(wrapper.vm.contentToHtml).toBe(announcementContentInHtml);
  });

  it("Should render the date in the correct format", () => {
    expect(wrapper.vm.date).toBe("December 15, 2022");
  });
});
