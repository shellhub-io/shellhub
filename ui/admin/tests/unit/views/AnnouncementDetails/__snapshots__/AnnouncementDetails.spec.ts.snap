// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Announcement Details > Renders the component 1`] = `
"<div class="d-flex pa-0 align-center">
  <h1>Announcement Details</h1>
</div>
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated mt-2 pa-4">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text">
    <div>
      <div class="text-overline mt-3">
        <h3>Uuid:</h3>
      </div>
      <div data-test="eac7e18d-7127-41ca-b68b-8242dfdbaf4c">
        <p>eac7e18d-7127-41ca-b68b-8242dfdbaf4c</p>
      </div>
    </div>
    <div>
      <div class="text-overline mt-3">
        <h3>Title:</h3>
      </div>
      <div data-test="Announcement 1">
        <p>Announcement 1</p>
      </div>
    </div>
    <div>
      <div class="text-overline mt-3">
        <h3>Content:</h3>
      </div>
      <div class="pa-2">
        <div data-test="## ShellHub new features 
 - New feature 1 
 - New feature 2 
 - New feature 3">
          <div class="content-announcement">
            <h2>ShellHub new features</h2>
            <ul>
              <li>New feature 1</li>
              <li>New feature 2</li>
              <li>New feature 3</li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div class="text-overline mt-3">
        <h3>Date:</h3>
      </div>
      <div data-test="2022-12-15T19:45:45.618Z">
        <p>December 15, 2022</p>
      </div>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
